### DISCUSSION TITLE
Simple application help

### DISCUSSION THREAD

**Discussion by Liu-jingyu-CN** [eModbus/eModbus#317] [Q&A] [answered]
Hello everyone, I now have an air flow meter that needs to obtain wind speed through modbus rtu, but since I am a newbie, I still haven’t obtained successful data after doing it for a long time. I hope you can help me.
1. The flow meter protocol is as follows
     send :  
                 server ID | function code | Register address | Data length | CRC
                 01  03  00 63   00 01  74 14
    return:
                 01  03   00 01   04 1A  CB 4F
   parse:  0x041A/100= 1050/100= 10.4m/s

2.  The hardware part uses mxa13487
    
![image](https://github.com/eModbus/eModbus/assets/103397361/d94718d4-2296-4f29-abce-dc5f17626767)

3. Software part,
This should need to be modified according to the needs, but my ability in this part is limited and I don't understand the examples. Can anyone help me make suggestions for corresponding modifications here?

`
#include <Arduino.h>
#include "ModbusClientRTU.h"
ModbusClientRTU MB;
void handleData(ModbusMessage response, uint32_t token) 
{
  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
  for (auto& byte : response) {
    Serial.printf("%02X ", byte);
  }
  Serial.println("");
}
void handleError(Error error, uint32_t token) 
{
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s\n", (int)me, (const char *)me);
}
void setup() {
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

  RTUutils::prepareHardwareSerial(Serial2);
  Serial2.begin(9600, SERIAL_8N1, GPIO_NUM_17, GPIO_NUM_16);

  MB.onDataHandler(&handleData);
  MB.onErrorHandler(&handleError);
  MB.setTimeout(2000);
  MB.begin(Serial2);

  uint32_t Token = 1111;

  Error err = MB.addRequest(Token++, 1, READ_HOLD_REGISTER, 0x0063, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

}

void loop() {
}
`
4. Serial port information
Error response: E2 - CRC check error
__ OK __
Error response: E2 - CRC check error
__ OK __
Error response: E2 - CRC check error
__ OK __
Error response: E2 - CRC check error
__ OK __
Error response: E2 - CRC check error
__ OK __
Error response: E2 - CRC check error

I hope you can help me, thank you

**Comment by Miq1** [comment]
Hi, I will try. 😉 

I staggered with the D3 diode in your schematics - would not that effectively prevent any signal on the RX line?

The request in the sketch looks good at first sight. The E2 CRC error seems to be responded to each request, so the first thing to check is the cabling. Is it really implemented as the schematic is describing? 

Next a full log will be interesting. If you know how, the addition of ``-DLOG_LEVEL=6`` to your compile flags will enable it. Alternatively you may edit the ``Logging.h`` source file and look for the line at the beginning that sets the default log level. Change that to ``LOGLEVEL_VERBOSE`` to get the same (do not forget to change it back at some time when you fixed your issue).

**Reply by Liu-jingyu-CN** [reply]
The circuit design is a reference to the available solutions, which can be used normally. In order to avoid this problem, I reused the product module for testing. The problem is the same. Now the log is as follows：

ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x43c
load:0x403c9700,len:0xbec
load:0x403cc700,len:0x2a3c
entry 0x403c98d8
__ OK __
[D] 99| ModbusClientRTU.cpp [ 111] setTimeout: Timeout set to 2000
[V] 99| RTUutils.cpp [ 112] calculateInterval: Calc interval(9600)=3645
[D] 110| ModbusClientRTU.cpp [ 86] doBegin: Client task 1070485920 started. Interval=3645
[V] 110| ModbusMessage.cpp [ 477] checkData: Check data https://github.com/eModbus/eModbus/issues/3
[D] 120| ModbusClientRTU.cpp [ 147] addRequestM: request for 01/03
[D] 130| ModbusClientRTU.cpp [ 230] addToQueue: RC=01
[D] 130| ModbusClientRTU.cpp [ 158] addRequestM: RC=00
[D] 209| ModbusClientRTU.cpp [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FCE5320/6:
| 0000: 01 03 00 63 00 01 |...c.. |
[D] 217| ModbusClientRTU.cpp [ 253] handleConnection: Request sent.
[V] 232| RTUutils.cpp [ 262] receive: C/3658us without data after 8
[V] 232| RTUutils.cpp [ 272] receive: C/[V] Raw buffer received: @3FCE537C/8:
| 0000: 7F 7E 39 FF FD 17 D7 00 |�~9..... |
[D] 252| RTUutils.cpp [ 437] receive: C/[D] Received packet: @3FCE5320/1:
| 0000: E2 |. |
[D] 263| ModbusClientRTU.cpp [ 268] handleConnection: Error response (1 bytes) received.
[V] Data: @3FCE5320/1:
| 0000: E2 |. |
[D] 273| ModbusClientRTU.cpp [ 289] handleConnection: Response generated.
[V] Response packet: @3FCE5320/3:
| 0000: 01 83 E2 |... |
Error response: E2 - CRC check error

How can I add CRC data to the sent message? Maybe this can solve the problem. It’s just my poor programming. I hope you can continue to help me.

**Reply by Miq1** [reply]
Forget about the CRC, it is added internally by the library. But as I was suspecting, something is very odd with the data on the line. 
This line
```
[V] Raw buffer received: @3FCE537C/8: | 0000: 7F 7E 39 FF FD 17 D7 00 |�~9..... |
```
is telling you what came back as a response. Your request `` 01 03 00 63 00 01 `` is asking for a single register, a proper response will start ``01 03 ...`` as well. The response you got has much too many ``1`` bits in it to make any sense.

I still would recommend to have a look at the wiring and just for testing purposes leave out that ominous D3 diode...

**Reply by Liu-jingyu-CN** [reply]
I'm glad that I found the problem with your help, but I can't understand it. After add one more 120R resistor to the bus, the work became normal. However, when the bus is 60R, it doesn't work properly.
Now I want to analyze the wind speed according to the protocol in the program. I still need your help. How do I need to modify the program to print out the wind speed correctly?

log:
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x43c
load:0x403c9700,len:0xbec
load:0x403cc700,len:0x2a3c
entry 0x403c98d8
__ OK __
[D] 99| ModbusClientRTU.cpp  [ 111] setTimeout: Timeout set to 2000
[V] 99| RTUutils.cpp         [ 112] calculateInterval: Calc interval(9600)=3645
[D] 109| ModbusClientRTU.cpp  [  86] doBegin: Client task 1070485920 started. Interval=3645
[V] 110| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 120| ModbusClientRTU.cpp  [ 147] addRequestM: request for 01/03
[D] 130| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 130| ModbusClientRTU.cpp  [ 158] addRequestM: RC=00
[D] 209| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FCE5320/6:
  | 0000: 01 03 00 63 00 01                                 |...c..          |
[D] 217| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[V] 296| RTUutils.cpp         [ 262] receive: C/3650us without data after 8
[V] 296| RTUutils.cpp         [ 272] receive: C/[V] Raw buffer received: @3FCE537C/8:
  | 0000: 01 03 02 00 00 B8 44 00                           |......D.        |
[D] 306| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FCE5334/6:
  | 0000: 01 03 02 00 00 B8                                 |......          |
[D] 327| ModbusClientRTU.cpp  [ 268] handleConnection: Data response (6 bytes) received.
[V] Data: @3FCE5334/6:
  | 0000: 01 03 02 00 00 B8                                 |......          |
[D] 337| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[V] Response packet: @3FCE5334/6:
  | 0000: 01 03 02 00 00 B8                                 |......          |
Response: serverID=1, FC=3, Token=00000457, length=6:
01 03 02 00 00 B8

**Reply by Miq1** [reply, marked as answer]
Regarding the 120 Ohms resistor: a separate one is required at each device connected to the bus. The two 10k biasing resistors have to be on the bus only once.

For the wind speed you will have to define two variables, one to read the register value from the response, one float to calculate the speed from it.

Once you got the ``response`` you could use some code like
```
uint16_t rawValue;
float windSpeed;
uint16_t offset = 3;
offset = response.get(offset, rawValue);
windSpeed = rawValue / 100.0;
```
Note that ``offset`` is used to set the byte in the response where to start extracting a value and is updated to the first byte behind with the return value of ``get()``.

**Reply by Miq1** [reply]
Please note that still something is wrong with your setup. there is a trailing zero byte ``0x00`` with the response. Thus the response is 8 bytes instead of the 7 bytes to be expected. I am amazed that this is passing the CRC check at all 😮

**Reply by Liu-jingyu-CN** [reply]
Thank you for your help. Now there is indeed one 0x00 byte missing. It can be used temporarily according to your help. I will provide the complete schematic and program as an answer later to facilitate everyone's better understanding in the future.

**Reply by Miq1** [reply]
No, there is a byte more than should be. In the response packet
```
[V] Raw buffer received: @3FCE537C/8: | 0000: 01 03 02 00 00 B8 44 00 |......D. |
```
there is ``01`` for the server ID, ``03`` for the function code, ``02`` as the length of data in bytes, ``00 00`` is the data proper and ``B8 44`` is the CRC. Thus the last ``00`` is extra.

This on RTU can happen with some RS485 adapters having floating pins when switching from listen to write mode or back, which can be interpreted as a zero byte if biasing or termination is not properly working. You could try with a 10k pulldown resistor at the RX wire to get a stable signal.

**Reply by Liu-jingyu-CN** [reply]
[空调监测.pdf](https://github.com/eModbus/eModbus/files/13307935/default.pdf)
Thank you again. After making the PCB, I verified that everything is normal. Thank you very much for your help.
